daemon/src/common/Profiler.cpp

```diff
@@ -0,0 +1,186 @@
+    * Neither the name of the Daemon developers nor the
+      names of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL DAEMON DEVELOPERS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+===========================================================================
+*/
+
+#include "Profiler.h"
+#include "../engine/framework/LogSystem.h"
+
+#include "../engine/framework/CommandSystem.h"
+#include "../engine/framework/BaseCommands.h"
+#include <fstream>
+
+
+namespace Profiler{
+
+    /*
+         * Function execution times are saved in the samples vector, every element
+         * contains the name of the function, the type of sample (function start,
+         * function end, and if there's a new frame), and a timestamp.
+         */
+    std::vector <Profiler::Point>                   samples;
+    std::chrono::high_resolution_clock::time_point  startTime;
+    bool                                            enabled    = false; ///TODO those bools are silly
+    bool                                            nextFrame  = false;
+    bool                                            shouldStop = false;
+
+
+    /// Called at the end of every frame
+    void Update(){
+
+        if(nextFrame){
+            enabled=true;
+            shouldStop=true;
+            nextFrame=false;
+            startTime   = std::chrono::high_resolution_clock::now();
+            return;
+        }
+
+        if(shouldStop){
+            enabled=false;
+            shouldStop=false;
+            Stop();
+            return;
+        }
+
+        if(!enabled)
+            return;
+        samples.push_back(Point(FRAME,""));
+    }
+
+
+    /// Point saves the time it's constructed
+    Point::Point(Type type,std::string label=""): type(type), label(label){
+        auto timeElapsed = std::chrono::high_resolution_clock::now() - startTime;
+        time = std::chrono::duration_cast < std::chrono::microseconds > (timeElapsed).count();
+    }
+
+
+    /// At the Beginning and End of every function
+
+    Profile::Profile( std::string label ): label( label ){
+        if(!enabled)
+            return;
+        samples.push_back( Point( START, label ) );
```

Comment:
Currently it'd need locking.
Locking would have a significant performance overhead but I doubt you can alter a vector lock-free